MoreLinq/GroupPartition.cs

```diff
@@ -0,0 +1,144 @@
+            Partition(source, k1, k2, null, resultSelector);
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching two
+        /// sets of keys and those groups that do not. An additional parameter
+        /// specifies how to compare keys for equality.
+        /// </summary>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey k1, TKey k2, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            PartitionImpl(source, 2, k1, k2, default(TKey), comparer, (a, b, c, rest) => resultSelector(a, b, rest));
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching three
+        /// sets of keys and those groups that do not.
+        /// </summary>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey k1, TKey k2, TKey k3,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            Partition(source, k1, k2, k3, null, resultSelector);
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching three
+        /// sets of keys and those groups that do not. An additional parameter
+        /// specifies how to compare keys for equality.
+        /// </summary>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey k1, TKey k2, TKey k3, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            PartitionImpl(source, 3, k1, k2, k3, comparer, resultSelector);
+
+        static TResult PartitionImpl<TKey, TElement, TResult>(IEnumerable<IGrouping<TKey, TElement>> source,
+            int count, TKey k1, TKey k2, TKey k3, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector)
+        {
+            Debug.Assert(count > 0 && count <= 3);
+
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+            List<IGrouping<TKey, TElement>> etc = null;
+
+            var groups = new[]
+            {
+                Enumerable.Empty<TElement>(),
+                Enumerable.Empty<TElement>(),
+                Enumerable.Empty<TElement>(),
+            };
```

Comment:
Why are the groups pre-inited to `Enumerable.Empty` but `etc` isn't? Any particular reason?